Extensions: Add extension reducers to static build

coderkevin opened the pull request:
This commit adds a reducer from any extention that has a
`state/reducer.js` file. It builds this in statically to the primary
reducer chain and runs it for all actions, just like any other reducer
in Calypso.

This is patterned off of PR #8103 from the original Calypso Extension
work.

It also adds a very simple example of using this state in the
`hello-dolly` example extension. Review of the example code was
performed already in PR #11347

To Test:
1. Run `make test` to verify all tests are passing.
2. Run `make run`
3. Visit `http://calypso.localhost:3000/hello-dolly`
4. Verify that the lyrics show up in bold.
5. Click 'Next' to see the next lyric.

dmsnell commented on client/state/index.js:
```diff
@@ -62,7 +63,7 @@ import wordads from './wordads/reducer';
  * Module variables
  */
-export const reducer = combineReducers( {
+const reducers = {
```
was it intentional to remove the export?

coderkevin replied:
Yep, so it could be placed below. No real reason to export the object mapping of original reducers in this case.

dmsnell commented on client/state/index.js:
```diff
@@ -111,7 +112,9 @@ export const reducer = combineReducers( {
-} );
+};
+
+export const reducer = combineReducers( { ...extensions.reducers(), ...reducers } );
```
we're making a choice here to overwrite extension reducers with the built-ins if an extension uses the same name as a reducer existing already in Calypso. is this the choice we want?

coderkevin replied:
Well, it was one or the other. I figured in the case of a name collision, it's better to not lose core functionality. Ideally, we'd namespace all extensions though, so namespacing issues like this can't happen.

gwwar replied:
>  Ideally, we'd namespace all extensions

Is there anything stopping us from doing this now? I would think we'd want this in before it's copy pasted anywhere else.

coderkevin replied:
Well, I was going off the original PR (#8103) as a baseline. However, I'd be up for something better here. We could automatically take care of this as part of the static build. I see two options:
1. Prefix all extension reducers. (e.g. `extension_{$extensionName}`)
2. Combine all extension reducers under a single one named `extensions`

gwwar replied:
I'm a fan of option 2 😄

coderkevin replied:
It should be impossible as the names of the extensions are derived from the listing of directories under `client/extensions`

coderkevin replied:
Combined extension reducers under `extensions` in the redux state tree.

dmsnell commented on server/bundler/extensions-loader.js:
```diff
@@ -0,0 +1,34 @@
+
+function getReducers( extensions = {} ) {
+	return extensions
+		.filter( ( extension ) => existsSync( path.join( 'client', 'extensions', extension, 'state', 'reducer.js' ) ) )
```
this applies a native filter and map over a value which according to the function declaration is an object. it _should_ throw a `TypeError`

was `extensions` supposed to default to `[]` instead?

coderkevin replied:
Yep. You're right. I'll correct it to default to `[]`

aduth replied:
Could we use an asynchronous `Promise.all` + `exists` + [`compact`](https://lodash.com/docs/4.17.4#compact) instead of `filter` + `existsSync`?

coderkevin replied:
Can we use async stuff like this in a webpack loader? I'm no expert here.

aduth replied:
>Can we use async stuff like this in a webpack loader? I'm no expert here.

Ah yeah, you have to opt in to it, but it's fairly straight-forward:

https://webpack.js.org/api/loaders/#async-loader

coderkevin replied:
Okay, Justin and I spent some time getting this in, but I think we've got it solved appropriately. Reviews definitely welcome with this new code here.

aduth replied:
>Okay, Justin and I spent some time getting this in, but I think we've got it solved appropriately. Reviews definitely welcome with this new code here.

Mmm... the logic here looks to me the same as it was before?

dmsnell commented on server/bundler/extensions-loader.js:
```diff
@@ -0,0 +1,34 @@
+}
+
+module.exports = function( content ) {
+	this.cacheable && this.cacheable();
```
what is this?

coderkevin replied:
It's a webpack cacheable thing. It's identical to the one in `server/bundler/loader.js`
Docs here: https://webpack.github.io/docs/how-to-write-a-loader.html#flag-itself-cacheable-if-possible

dmsnell commented on server/bundler/extensions-loader.js:
```diff
@@ -0,0 +1,34 @@
+module.exports = function( content ) {
+	this.cacheable && this.cacheable();
+
+	const extensions = require( this.resourcePath );
```
where did these `this.` values come from?

aduth replied:
>where did these `this.` values come from?

These are provided by Webpack to loaders.

https://webpack.js.org/api/loaders/#the-loader-context

gwwar commented on client/extensions/hello-dolly/state/reducer.js:
```diff
@@ -0,0 +1,25 @@
+}
+
+function advanceToNextLyric( index ) {
+	return index + 1;
```
While unlikely index will eventually overflow. Better to mod here than in the selector.

coderkevin replied:
Hmm, if we're really worried about this, I guess we'd need the length of the lyrics we're using, but that's not really accessible here. We could put the length in as state, but if we do, I wonder if we might as well just put all the lyrics in as state then.

dmsnell replied:
hearkening back to earlier comments I have made: I'm not entirely sure we've settled on the most appropriate data structure and this is one of the indicators. we have mutually dependent state in separate places that easily gets out of sync with each other

gwwar replied:
It wouldn't be anything too crazy. The lyrics are a constant anyway. I just found it a bit silly to find such a high index value in the state tree.

```
import lyrics from './lyrics';
//...
function advanceToNextLyric( index ) {
    return ( index + 1 ) % lyrics.length; 
}
```

dmsnell replied:
it's not too silly really @gwwar since it's a safe operation to keep on increasing. it's also safe against certain kinds of bugs where the lyrics change

in general I'm hesitant to store lists and indices separate. one of my suggestions was to store the whole list here and rotate it instead of using an index. I like keeping the lyrics length out of this file. it might be that we picked a fairly difficult data type for creating an example

gwwar replied:
> in general I'm hesitant to store lists and indices separate. one of my suggestions was to store the whole list here and rotate it instead of using an index.

That seems fine to me. tbh this is a very trivial example.

As a plugin developer the more interesting question would be, where do I store my session data vs long term data and what does that look like?

mcsf replied:
I agree with @gwwar. Let's keep this focused on the PR's intent, not the example, which has been debated at length in the parent PR.

mtias replied:
I'd be ok with `hello-dolly` not adding anything to state for now as well and only using existing state through selectors to keep it simpler.

Not a strong opinion here, but it may be wiser to do the lyrics example once we figure out how to dynamically load this stuff.

coderkevin replied:
Added a mod check for `lyrics.length`. I think this is logical enough since we're using the default export that implies the static lyrics. The other export in the selector is used for tests only.

mcsf replied:
Pro tip: `hello-dolly` could work without similarly to how it works right now but without its own reducer if it consumed `state.ui.actionLog` — see Guided Tours' [`ARCHITECTURE.md`](https://github.com/Automattic/wp-calypso/blob/master/client/layout/guided-tours/docs/ARCHITECTURE.md).

mcsf commented on client/extensions/hello-dolly/index.js:
```diff
@@ -7,52 +7,13 @@ import React from 'react';
-			</FAQ>
-		</Main>
-	), document.getElementById( 'primary' ), context.store );
+	renderWithReduxStore( <HelloDollyPage />, document.getElementById( 'primary' ), context.store );
```
For brevity, we could use [`renderPage`](https://github.com/Automattic/wp-calypso/blob/442c64b5b4ae97df858e688fae6893f1fa1d817c/client/lib/react-helpers/index.js#L17-L23).

coderkevin replied:
Thanks for the tip!

mcsf commented on client/extensions/hello-dolly/state/selectors.js:
```diff
@@ -0,0 +1,19 @@
+ */
+import lyrics from './lyrics';
+
+export const getCurrentLyric = ( lines ) => ( reduxState ) => {
```
Minor: why not just call the variable `state`? We're in a file of path `state/selectors.js`, so I don't think there will be doubts as to whether we're talking about Redux state or React component state.

coderkevin replied:
The ambiguity here is whether or not we're passing the entire redux state tree or a sub-node of that state that applies to this reducer. For instance, the reducer gets passed the sub-node, but this selector expects the top-level redux state tree.

I do feel we'll have to figure this out at a much larger level at some point. Personally, I'm not a big fan of passing around the entire redux state tree all the time, but that's a much larger discussion.

mcsf commented on client/extensions/hello-dolly/state/test/reducer.js:
```diff
@@ -0,0 +1,27 @@
+		expect( reducer( 11, { type: ROUTE_SET } ) ).to.equal( 12 );
+		expect( reducer( 12, { type: SECTION_SET } ) ).to.equal( 13 );
+		expect( reducer( 13, { type: SITE_SETTINGS_SAVE } ) ).to.equal( 14 );
+	} );
```
Maybe add a bogus action to make sure it doesn't advance? Disregard if too paranoiac.

coderkevin replied:
Added test for bogus action.

coderkevin commented:
Updates:

- Added test for bogus action to not increment.
- Added mod check for `lyrics.length` in reducer.
- Using `renderPage` in `index.js`
- Combined extension reducers under extensions in the redux state tree.
- Went async on the extension reducer code generation.

We have an issue with running tests with this PR and could use some help webpack-wise with it. Inside `client/state/index.js`, when it imports `extensions`, it gets the original file instead of the one transformed by `server/bundler/extensions-loader`. This only happens on `make tests`. It runs fine normally. @ehg, @aduth or someone more familiar with this, could you help?

aduth commented:
>Inside `client/state/index.js`, when it imports `extensions`, it gets the original file instead of the one transformed by `server/bundler/extensions-loader`.

Oof, yeah, we've hit this a few times when trying to introduce new loaders. Basically the problem is that Mocha is only run with Babel as a compiler; Webpack doesn't enter the picture at all. You can see some previous discussion where I struggled with this for a custom `async-component` loader at https://github.com/Automattic/wp-calypso/pull/8544#discussion_r82464155 . In the end, we decided to leave out the loader.

In order to make it work, it seems we might need to explore running Mocha as part of the Webpack pipeline or at least after Webpack has had a chance to apply its transformations. Admittedly I didn't have much luck with this when I tried, so I don't have a ton of advice to offer.

coderkevin commented:
Thanks for the insight, @aduth. Maybe we could just manually run the loader outside of webpack and point mocha to it?

gziolo commented on server/bundler/extensions-loader.js:
```diff
@@ -0,0 +1,74 @@
+}
+
+function generateExtensionsModuleString( reducerRequires ) {
+	return `module.exports = {
```
Could it have the following code instead?

```js
return `
const combineReducers = require( ''redux' ).combineReducers;

module.exports = combineReducers(
    ${ reducerRequires }
);`;

Then it `client/state/index.js` it would be enough to have the following import statement:
```js
import extensions from 'extensions';
```

coderkevin replied:
I suppose it could. I'll think about this, especially in the context of the testing issue for this PR as well.

gziolo commented:
Nice work on this one 👍

coderkevin commented:
So, it looks like the failing tests are the only thing that is holding this PR up at this point. However, the reason the tests aren't passing is a larger one which needs discussion outside this PR. So I'd like to have a discussion here on how to move this PR forward. Here's the problem:

Because `state/index` now imports `client/extensions`, which is a webpack-translated file in normal builds, but not in test.

So, the way I see it, we have 3 options ahead of us:
1. Temporarily remove the `client/state/test/index` tests until we get this sorted (not very desirable)
2. Modify `client/state/index` to only import `client/extensions` on dev/prod build and not test builds.
3. Add special code to `client/extensions` to work in a node/test build without the use of the webpack translate and return the same data structure that is created by the loader.
4. Make a big change to tests and wait for that to be done before this gets in (blocks the WooCommerce project until we get this done)

Thoughts?

gziolo commented:
You can try this helper https://github.com/Automattic/wp-calypso/blob/master/test/test/helpers/use-filesystem-mocks/README.md. See readme and usages in our codebase :)

coderkevin commented:
Okay, @gziolo, I think I see what you're getting at. You're saying to create a mock `client/extensions/index.js` file that I would load in with `use-filesystem-mocks` that would then be loaded by `client/state/index.js` when I include that in my tests? I would need to change my `import` to a `require` to make that work, right?

gziolo commented:
Yes, this is how it would work.

Another solution would be to put mock file in /test/extensions/index.js. This is hacky solution that would take advantage of how NODE_PATH is defined for tests :)

coderkevin commented:
Okay, @gziolo (and others), I tried `use-filesystem-mocks` for my tests, and it worked. However, there are 7 other tests that import from `'state'`. So is it best to mock out `'state'` for each of those? Or put the mock file in `/test/extensions/index.js`. Either way, I think we should consider this a temporary fix until #12203 is addressed.

ehg commented:
> Another solution would be to put mock file in /test/extensions/index.js. This is hacky solution that would take advantage of how NODE_PATH is defined for tests :)

I have no problem with this approach, until we have a better one. Let's do this for now, include a `TODO` comment, and get this PR in.

coderkevin commented:
@dmsnell, @mtias, or any others, are we good to merge this now?

mcsf commented on client/extensions/hello-dolly/hello-dolly-page.js:
```diff
@@ -0,0 +1,102 @@
+	);
+}
+
+export default connect( mapStateToProps, mapDispatchToProps )( HelloDollyPage );
```
Are the `map*ToProps` laid out so verbosely for a specific reason? For illustrative purposes? I'd argue that they make Redux seem more intimidating to unaccustomed eyes. The naming, though standard, is arguably intimidating, but the general overhead and the use of `bindActionCreators` could also be avoided. Compare with:

```js
export default connect(
  ( state ) => ( {
    currentLyric: getLyric( state )
  } ),
  {
    nextLyric,
  }
)( HelloDollyPage );
```

Not a blocker, but I'd aim to keep things simple.

dmsnell replied:
I agree with staying close to the docs @coderkevin but I'm also willing to diverge knowing that the examples are sometimes given in styles that are specifically geared towards being examples and not production code.

Inlining the functions irks me @mcsf and is confusing because it's harder for me to grok what is happening where, plus those functions are just meshed inside of the function call

On the other hand, I like very simple mapping functions and think we often make them needlessly verbose

```js
const mapStateToProps = state => ( {
	currentLyric: getLyric( state ),
} );

const mapDispatchToProps = {
	nextLyric,
};
```

having separate functions for these is definitely a very frequent pattern in Calypso but we haven't chosen one-pattern-to-rule-them-all because we disagree on what is most appropriate.

mcsf replied:
(Side thought: `mapActionsToProps` would be much more adequate for the kind of (auto-)binding we perform in `mapDispatchToProps`.)

dmsnell replied:
@mcsf that would be the case in some circumstances but in others it really is a mapping from dispatch + other stuff into props

mcsf replied:
I think we're agreeing on the same thing. :)

mcsf commented on server/bundler/extensions-loader.js:
```diff
@@ -0,0 +1,74 @@
+	};`;
+}
+
+function getReducersSync( extensionDirs = {} ) {
```
This defaults to `{}` instead of `[]`.

mcsf commented on server/bundler/extensions-loader.js:
```diff
@@ -0,0 +1,74 @@
+	return `module.exports = {
+		reducers: function() {
+			return {
+				${ reducerRequires }
```
I know that the VM takes care of joining the array's lines when interpolating into the string template, but I'd play it safe and explicitly join:
```js
`
			return {
				${ reducerRequires.join( separator ) }
`
```

If you choose `separator := '\n'`, you can ditch the newline in `generateReducerRequireString`.

mcsf commented on server/bundler/extensions-loader.js:
```diff
@@ -0,0 +1,74 @@
+			} )
+		).then( ( reducerRequires ) => resolve( compact( reducerRequires ) ) );
+	} );
+}
```
Minor, but there's lexical nesting with `resolve` referring to different things depending the line. Suggestion: `resolveAll` and `resolveOne`?

mcsf commented on server/bundler/extensions-loader.js:
```diff
@@ -0,0 +1,74 @@
+		getReducersAsync( extensionDirs )
+			.then( ( reducerRequires ) => resolve( generateExtensionsModuleString( reducerRequires ) ) );
+	} );
+}
```
From the looks of it, is the `new Promise` construct necessary? We already grab a promise from `getReducersAsync`, so it would be:

```js
function getExtensionsModuleAsync( extensionDirs ) {
    return getReducersAsync( extensionDirs )
        .then( generateExtensionsModuleString );
}
```

or am I missing something?

dmsnell replied:
thenables which return plain values can continue the Promise chain to the next thenable without explicitly creating a new Promise

coderkevin commented:
Addressed comments by @mcsf

coderkevin commented:
>Let's make sure all of the e2e tests run before deploying.

How's the best way to do this, @dmsnell? I started down the road of running them all locally, but it looks like it needs a lot of config data.

dmsnell commented:
> How's the best way to do this, @dmsnell? I started down the road of running them all locally, but it looks like it needs a lot of config data.

calling in the help of @alisterscott or @stevekaeser for this one. I'm a n00b on our e2e tests.

with some work that @samouri did too we scheduled an all-hands-on-deck deployment where people around Calypso all tested out their own things they had worked on to manually verify the tests. this was for a PR with similar possible implications and leverage as this one. in my opinion it worked out well and we all shared the responsibility for the deploy. I would love to see this happen again.

alisterscott commented:
I ran all the e2e tests against this PR.

All e2e tests passed besides the customizer tests which failed locally on my machine as the customizer wasn't displaying the site, but I also tried master and the same problem appeared so I don't think it's specific to this PR. Customizer also loads via https://calypso.live/?branch=add/extensions-static-reducers, so 👍

ehg commented:
> this was for a PR with similar possible implications and leverage as this one. in my opinion it worked out well and we all shared the responsibility for the deploy. I would love to see this happen again.

I think the implications are less dangerous than the single-build PR :) But yes, when you merge this @coderkevin — give us a heads up in #calypso

mcsf commented on server/bundler/extensions-loader.js:
```diff
@@ -0,0 +1,72 @@
+function getReducersSync( extensionDirs = [] ) {
+	return extensionDirs
+		.filter( ( extensionDir ) => fs.existsSync( getReducerPath( extensionDir ) ) )
+		.map( ( extensionDir ) => generateReducerRequireString( extensionDir ) );
```
The above could be just `.map( generateReducerRequireString );`, but totally not a blocker.

mcsf commented on server/bundler/extensions-loader.js:
```diff
@@ -0,0 +1,72 @@
+			} )
+		).then( ( reducerRequires ) => resolveAll( compact( reducerRequires ) ) );
+	} );
+}
```
Again, not a blocker, but I'm realizing that

```js
new Promise( resolveAll ) =>
  Promise.all( promises ).then( requires => resolveAll( compact( requires ) ) )
```

is probably equivalent to

```js
Promise.all( promises ).then( compact )
```

Worth looking into after this gets merged — less logic means easier to maintain code, and it's easier to refactor when it's still fresh in our minds.

mcsf commented:
@coderkevin: any items left on this one before merging? Anything we can we help with?

coderkevin commented:
I think we're ready to go on this. I had intended to merge it yesterday but ended up being out sick. I'll merge this today now that I can keep an eye on it. :)
